### PR TITLE
[ESIMD] Deprecate -fsycl-explicit-simd option

### PR DESCRIPTION
**Pull request by DenisBakhvalov**
ESIMD no longer requires -fsycl-explicit-simd flag.

### CODE REVIEW

**Comment by DenisBakhvalov** [comment]
Documentation will be updated in a separate patch. I'm working on it right now.

**Comment by DenisBakhvalov** [comment]
> Documentation will be updated in a separate patch. I'm working on it right now.

The patch is here: https://github.com/intel/llvm/pull/3477

**Review by mdtoguchi — COMMENTED** [review]
Are there any users of the option?  Does this need to go through a deprecation process before we remove?

**Comment by DenisBakhvalov** [comment]
> Are there any users of the option? 
You mean any external users, right?

> Does this need to go through a deprecation process before we remove?
ESIMD is still an experimental feature, however, I'm not sure about this. @kbobrovs, what do you think?

**Comment by AaronBallman** [comment]
> > Are there any users of the option?
> > You mean any external users, right?
> 
> > Does this need to go through a deprecation process before we remove?
> > ESIMD is still an experimental feature, however, I'm not sure about this. @kbobrovs, what do you think?

How long as the experimental feature been available for to users? Do we know of anyone using it?

**Comment by kbobrovs** [comment]
Yes, this is experimental feature. But I suggest to leave it there, adding a deprecation message and saying that it has no effect now.
+ @pvchupin

**Comment by pvchupin** [comment]
> Yes, this is experimental feature. But I suggest to leave it there, adding a deprecation message and saying that it has no effect now.
> 
> * @pvchupin

I think it's OK to update all our tests reflecting new behavior. For the option itself we need a message that it's deprecated and not needed anymore.

**Comment by pvchupin** [comment]
Also please add it to the list at https://github.com/intel/llvm/blob/sycl/sycl/doc/UsersManual.md under [DEPRECATED] tag

**Comment by DenisBakhvalov** [comment]
> > Yes, this is experimental feature. But I suggest to leave it there, adding a deprecation message and saying that it has no effect now.
> > 
> > * @pvchupin
> 
> I think it's OK to update all our tests reflecting new behavior. For the option itself we need a message that it's deprecated and not needed anymore.

Ok, will update the patch.

**Review by pvchupin — APPROVED** [review]
ok for doc

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:5411]: The option doesn't look to be used anymore, so would it be better to modify the option in `Options.td` to be part of the `clang_ignored_legacy_options_Group` group?

Comment (DenisBakhvalov) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Right.

File: clang/include/clang/Driver/Options.td
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:4378]: Thanks for updating here - now that the deprecated message comes out and the option isn't hooked to anything, stating that the option does anything is misleading.  @pvchupin, how would be best to represent the information shown in the help?  Maybe something like `Deprecated - enabling explicit SIMD extensions is on by default`

File: sycl/doc/UsersManual.md
Comment (kbobrovs) [inline sycl/doc/UsersManual.md:58]: It rather has not effect now.

**Review by alexbatashev — APPROVED** [review]
Test changes LGTM

File: sycl/doc/UsersManual.md
Comment (DenisBakhvalov) [reply sycl/doc/UsersManual.md]: Fixed.

**Review by mdtoguchi — APPROVED** [review]

File: clang/include/clang/Driver/Options.td
Comment (DenisBakhvalov) [reply clang/include/clang/Driver/Options.td]: Fixed. Now it says:
```
  -fsycl-explicit-simd    Enable SYCL explicit SIMD extension. (deprecated)
  -fno-sycl-explicit-simd Disable SYCL explicit SIMD extension. (deprecated)
```

File: clang/include/clang/Driver/Options.td
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:4377]: Without `CoreOption` the option will not be available for Windows.

Comment (DenisBakhvalov) [reply clang/include/clang/Driver/Options.td:4377]: Windows CI job seemed to pass.
Added CoreOption flag.

**Review by alexbatashev — APPROVED** [review]

**Review by kbobrovs — APPROVED** [review]

File: clang/include/clang/Driver/Options.td
Comment (AaronBallman) [inline clang/include/clang/Driver/Options.td:4373]:
```suggestion
// FIXME: -fsycl-explicit-simd is deprecated. remove it when support is dropped.
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a tiny commenting nit.

**Comment by DenisBakhvalov** [comment]
@bader, I believe some of the reviewers are on vacation. Can we consider this patch as approved?

**Comment by bader** [comment]
> @bader, I believe some of the reviewers are on vacation. Can we consider this patch as approved?

You need an approval from driver code owners. Even if they are on vacation (which according to my information is not the case), I don't understand why this patch can be considered as approved. I see that Mike left multiple comments, so I would like to get his approval before merging the patch.

**Review by mdtoguchi — APPROVED** [review]

**Pull request merged**
